torch/distributed/_functional_collectives.py

```diff
@@ -625,8 +662,13 @@ def wrap(e: torch.Tensor):
         unwrapped_args = tree_map_only(AsyncCollectiveTensor, unwrap, args)
         unwrapped_kwargs = tree_map_only(AsyncCollectiveTensor, unwrap, kwargs)
 
-        # we don't wrap the result as it doesn't need to be waited on.
-        out = func(*unwrapped_args, **unwrapped_kwargs)
+        # If elem has a grad_fn we need to enable reentrant_dispatch so that
+        # grad is propagated correctly. Normally the grad_fn would be on
+        # AsyncCollectiveTensor but since we wrapped a tensor with a grad_fn we
+        # need this.
+        with torch.overrides.enable_reentrant_dispatch():
```

Comment:
cc @zou3519 how do you feel about this one?
In this case, the wrapper doesn't require grad or has a grad_fn, only the wrapped Tensor and so we run autograd on the underlying Tensor here.
Off the top of my head, I think this will go wrong:
- When mixed with other subclasses that are resolved earlier (like DTensor). Where they will run their autograd on the wrapper (thinking it doesn't require grad) and we will create two different autograd graph
- If anyone checks the requires_grad field of the wrapper when it is returned (like a custom autograd.Function), it will also return the wrong thing.

The proper fix here is to fix the constructor to use a differentiable constructor like is done for other wrapper subclasses. For example DTensor: https://github.com/pytorch/pytorch/blob/f61b04a1f0ed55aa3f9b75e1266e7a5dc71fc90d/torch/distributed/_tensor/api.py#L37

Reply:
Yeah we don't want `enable_reentrant_dispatch` here

Reply:
Thanks for pointing out those wrappers -- I was looking for a way to do this but couldn't find anything